components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,416 @@
+    const std::vector<Log>& logs,
+    base::Value rawlogs,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::ProviderError::kSuccess) {
+    if (tasks.find(task_id) != tasks.end()) {
+      tasks[task_id]->MarkComplete();
+      MergeAllResultsAnCallBack();
+    }
+    return;
+  }
+
+  std::vector<Log> logs_tmp{logs};
+  std::sort(logs_tmp.begin(), logs_tmp.end(),
+            [](const Log& first, const Log& second) -> bool {
+              return std::tie(first.block_number, first.log_index) <
+                     std::tie(second.block_number, second.log_index);
+            });
+
+  // Collection of the latest allowances per contract & spender & approver.
+  std::map<std::string, std::tuple<uint256_t, mojom::AllowanceInfoPtr>>
+      allowance_map;
+  // Put cached data into the map first if available.
+  LoadCachedAllowances(chain_id, hex_account_address, allowance_map);
+  for (const auto& log_item : logs_tmp) {
+    // Skip pending logs.
+    if (log_item.block_number == 0) {
+      continue;
+    }
+
+    if (log_item.topics.size() < kTopicElementsCountToCheck) {
+      continue;
+    }
+    const auto current_map_key = GetAllowanceMapKey(
+        log_item.address, log_item.topics[1], log_item.topics[2]);
+
+    uint256_t parsed_amount{0};
+    if (!HexValueToUint256(log_item.data, &parsed_amount)) {
+      continue;
+    }
+
+    if (parsed_amount > 0) {
+      // Replace if same key exists by the fresh allowance data.
+      allowance_map.insert_or_assign(
+          current_map_key,
+          std::make_tuple(log_item.block_number,
+                          mojom::AllowanceInfo::New(
+                              chain_id, log_item.address, log_item.topics[1],
+                              log_item.topics[2], log_item.data)));
+    } else if (allowance_map.find(current_map_key) != allowance_map.end()) {
+      allowance_map.erase(current_map_key);
+    }
+  }
+  std::vector<mojom::AllowanceInfoPtr> allowances_found;
+  uint256_t max_block_number{0};
+  for (const auto& [allowance_map_key, allowance_tuple] : allowance_map) {
+    allowances_found.push_back(std::get<1>(allowance_tuple).Clone());
+    if (std::get<0>(allowance_tuple) > max_block_number) {
+      max_block_number = std::get<0>(allowance_tuple);
+    }
+  }
+  if (tasks.find(task_id) != tasks.end()) {
+    tasks[task_id]->SetResults(chain_id, max_block_number,
+                               std::move(allowances_found));
+    MergeAllResultsAnCallBack();
+  }
+}
+
+bool EthAllowanceManager::IsAllTasksCompleted() {
+  return !discover_eth_allowance_callback_.empty() &&
+         tasks.end() ==
+             std::find_if(
+                 tasks.begin(), tasks.end(),
+                 [](std::pair<
+                     const int,
+                     std::unique_ptr<EthAllowanceManager::EthAllowanceTask>>&
+                        item) { return !(item.second->is_completed); });
+}
```

Comment:
1) This should never be called in context of empty discover_eth_allowance_callback_ list, so `DCHECK(!discover_eth_allowance_callback_.empty())`
2) Also instead of find_if that should be done via base::ranges::all_of like
```
return base::ranges::all_of(
    tasks, [](auto& item) { return item.second->is_completed; });
```
3) Please mark method as const